test/Yesod/Paginator/WidgetsSpec.hs

```diff
@@ -189,3 +188,21 @@ spec = withApp $ do
             , "<li class=\"next disabled\"><a>»</a></li>"
             , "</ul>"
             ]
+
+    describe "filterParams" $ it "works" $ do
```

Comment:
I might flesh this out a bit more,

```hs
describe "filterParams" $ do
  it "inserts page number into params that lack it"

  it "updates an existing page number"

  it "doesn't unique other duplicated parameters"
```

WDYT?

Reply:
Oh!

You don't need to do this, because it could be annoying if you're not familiar with QuickCheck, but this seems like a great case for property-based testing. Things "it inserts or updates...", "it preserves order", or "it doesn't add or remove not-ours elements" could be expressed as properties and [quick-]checked on an arbitrary list of parameters.

Reply:
Haven't used property-based testing in practice but I think it sounds like a good idea, I'll see if I can come up with something!

Reply:
Regarding the property "it preserves order". As implemented, if the page parameter has the same name as a key of an item in the array, the item in the array will get filtered out, i.e.:
```
> filterParams "p" [("p", "foo")]
[("p", 1)]
```
I'm not sure whether:
- the implementation is incorrect
- we should generate data s.t. this doesn't occur in the test
- this behavior should be expressed in the property, if possible

What do you think?

Reply:
That's correct and desired. If you navigate to page 1, the URL will include `?p=1`; if you navigate to page 2 from there, it must _replace_ that with `?p=2`. That's why I've been calling this function append-or-replace -- it's not strictly append and not strictly replace, it needs to do both as necessary. I guess this operation on maps is called "insert", so maybe we can use that name.

Reply:
Yeah, I understand and agree about replacing the pagination parameter. I'm unsure about how that works out with the "it preserves order" property though:
```
buildParams "p" [("n", "foo"), ("p", "bar")]
> [("p", 1), ("n", "foo)]
```
I.e. since the pagination parameter is prepended, the order isn't preserved. Do you think we should drop or relax that property?